### ISSUE TITLE
Core dumped when using autotune

### ISSUE THREAD

**Issue by dveselov** [facebookresearch/fastText#888]
Hello, thanks for new autotune feature.
I'm trying to test it on my data, but got this exception:
```
$ fasttext supervised -input fasttext_train.txt -output fasttext_model -autotune-validation fasttext_valid.txt

Progress:  14.0% Trials:    6 Best score:  0.647911 ETA:   0h 4m17s

terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted (core dumped)
```
I'm attaching my train data, so you can reproduce this bug and is there anything info that I can provide also?
Thanks.

[fasttext_valid.txt](https://github.com/facebookresearch/fastText/files/3545921/fasttext_valid.txt)
[fasttext_train.txt](https://github.com/facebookresearch/fastText/files/3545922/fasttext_train.txt)

**Comment by Celebio** [comment]
Hi @dveselov ,
Thank you for reporting the issue. We will have a look.

Regards,
Onur

**Comment by dveselov** [comment]
Most likely that was caused by insufficient memory, because with `-wordNgrams 1` flag everything ends successfully.

**Comment by arhipovdv** [comment]
have same error with autotune, VM with 6 vCPU and 6Gb RAM

**Comment by kumar-shridhar** [comment]
For me `-autotune-validation` is generating an error: `Unknown argument: -autotune-validation`. I am using the C++ version.

**Comment by Celebio** [comment]
I can't reproduce on my machines. I am also guessing a memory issue depending on the explored parameters and the machine itself. I will continue my investigations.

**Comment by Celebio** [comment]
@dveselov , @arhipovdv , can you please try those :

- in `autotune.cc`, place loggers right before the training : 
```
    LOG_VAL(Trial, trials_)
    printArgs(trainArgs, autotuneArgs);
```
just before `fastText_->train(trainArgs);`
- compile and launch your executable with `-verbose 4`

and check if it actually tries a much bigger model right before crashing.

Can you also send the core file and your executable (as well as the platform details) ?

**Comment by arhipovdv** [comment]
> @dveselov , @arhipovdv , can you please try those :
> 
> * in `autotune.cc`, place loggers right before the training :
> 
> ```
>     LOG_VAL(Trial, trials_)
>     printArgs(trainArgs, autotuneArgs);
> ```
> 
> just before `fastText_->train(trainArgs);`
> 
> * compile and launch your executable with `-verbose 4`
> 
> and check if it actually tries a much bigger model right before crashing.
> 
> Can you also send the core file and your executable (as well as the platform details) ?
output with -verbose 4:
[log.txt](https://github.com/facebookresearch/fastText/files/3549373/log.txt)
executable:
[fasttext.zip](https://github.com/facebookresearch/fastText/files/3549375/fasttext.zip)

core file - what is it? where to find?

Ubuntu 19.04, VM under Hyper-V, 6 vCPU, 2/6/10 Gb RAM

**Comment by DanzaMacabra** [comment]
**System Information:**
- OS Platform and Distribution
Ubuntu 16.04 
- Python version:
Python 3.6.8
- GCC version:
GCC 8.0.1 20180414 (experimental) [trunk revision 259383]] on linux
- CPU model and memory:
Intel(R) Core(TM) i5-8500 CPU @ 3.00GHz
RAM 15.6G

I ran into similar problem when using python api(build from source return error 1). The RAM usage spike and PC freezed.
By adding logs args before autotune.cc, it did try loading a large model before crashing:
Trial = 10
epoch = 100
lr = 0.0219047
dim = 657
minCount = 1
wordNgrams = 3
minn = 0
maxn = 0
bucket = 10000000
dsub = 2
loss = hs 
Since I was forced to restart PC, there's no crash file in /var/crash

**Comment by lucaventurini** [comment]
Hi @Celebio !

I also had experienced this. In my case, it fails when it tries to set the dsub parameter to 16. By the way, I am a bit confused by this parameter, since it is listed in the quantization options in the docs but I have not set a size constraint. What is it doing?

**Comment by lucaventurini** [comment]
If I set manually dsub to 2, it seems working. But If I try to set other options together with autotuning, like -epoch 50 or -bucket 200000 , I get an unspecified "floating point exception", after the first trial, probably during the second.

**Comment by sisiruowan** [comment]
> If I set manually dsub to 2, it seems working. But If I try to set other options together with autotuning, like -epoch 50 or -bucket 200000 , I get an unspecified "floating point exception", after the first trial, probably during the second.

I have the same problem.

**Comment by OpenWaygate** [comment]
Same problem, anyone know how to solve it? 
```
Progress:  10.0% Trials:    5 Best score:  0.608669 ETA:   8h59m42s
terminate called after throwing an instance of 'std::bad_alloc'
what():  std::bad_alloc
```

**Comment by Celebio** [comment]
Hi @dveselov, @arhipovdv , @DanzaMacabra , @lucaventurini , @sisiruowan , @OpenWaygate ,
Could you please try at the commit https://github.com/facebookresearch/fastText/commit/a54ce12ccc3b5f37fa3758958d946e43f63969f8 ?

Best regards,
Onur

**Comment by TechnikEmpire** [comment]
Same issue at head:master with 32 gigs of ram, the application doesn't even approach 4% of available memory yet appears to fail to alloc and then tries to access the resulting null pointer.

```
Faulting application name: fasttext-latest.exe, version: 0.0.0.0, time stamp: 0x600686f3
Faulting module name: ucrtbased.dll, version: 10.0.19041.1, time stamp: 0xe7caee08
Exception code: 0xc0000005
Fault offset: 0x0000000000156210
```

**Comment by TechnikEmpire** [comment]
It's a result of providing args that autotune wants to work out itself.

```
fasttext-latest.exe supervised -thread 7 -input Data-Refined\eng_train.txt.train -output Models\out_file -loss ova -autotune-validation Data\eng_train.txt.validation -autotune-modelsize 6M -autotune-duration 600
Warning : loss is manually set to a specific value. It will not be automatically optimized.
```

If I remove any params that are warned about (like ova loss), then this crash doesn't happen.

**Comment by hafiz031** [comment]
If I am not mistaken, this happens due to the `bucket_size`, the auto tuner sets the `bucket_size` too large value (without considering if the machine will be able to handle it or not) and which overflows the memory. In may case it was set to `5056467` at some point. But I couldn't find a way to prevent `fastText autotune` from doing so.

**Comment by endeesa** [comment]
In my case the error went away after I restricted the model size using the autotuneModelSize argument. 

```python
fasttext.train_supervised(input=dataset.train_path.as_posix(),
                                           autotuneValidationFile=dataset.valid_path.as_posix(),
                                           autotuneDuration=duration_in_minutes*60,
                                           autotuneModelSize="2M")
```

Platform: Ubuntu 20.04 running on WSL2
VM memory: ~6GB
